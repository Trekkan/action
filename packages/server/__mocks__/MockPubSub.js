import dataFieldsToSerialize from '../__tests__/utils/dataFieldsToSerialize'
import * as getPubSub from '../utils/getPubSub'

const fieldsToSerialize = {
  invitation: [...dataFieldsToSerialize],
  newAuthToken: ['channelId', 'message.data.tms'],
  notification: [...dataFieldsToSerialize],
  organization: [...dataFieldsToSerialize],
  task: [...dataFieldsToSerialize],
  team: [...dataFieldsToSerialize],
  teamMember: [...dataFieldsToSerialize]
}

const untestablePayloads = ['message.data.removedRequestNotifications']

// const getPath = (str, obj) => str.split('.').reduce((o, i) => o[i], obj);
const replacePath = (str, docArr, newVal) => {
  docArr.forEach((obj) => {
    const path = str.split('.')
    let currentObj = obj
    for (let ii = 0; ii < path.length - 1; ii++) {
      const next = path[ii]
      currentObj = currentObj[next]
      if (!currentObj) return
    }
    const lastVal = path[path.length - 1]
    if (Array.isArray(currentObj[lastVal])) {
      currentObj[lastVal] = newVal
    }
  })
}

export default class MockPubSub {
  constructor () {
    getPubSub.default = () => this
    this.db = {}
  }

  __serialize (dynamicSerializer) {
    const channels = Object.keys(this.db).sort()
    const snapshot = {}
    for (let i = 0; i < channels.length; i++) {
      const channel = channels[i]
      const doc = this.db[channel]
      const channelFields = fieldsToSerialize[channel]
      if (!channelFields) {
        throw new Error(`BAD MOCK: No fieldsToSerialize for pubsub channel ${channel}`)
      }

      // some payloads can't be tested because they are an array that can't be sorted
      // for those, we can skip it here and use the arrayContaining test
      untestablePayloads.forEach((untestable) => {
        replacePath(untestable, doc, 'UNTESTED')
      })
      snapshot[channel] = dynamicSerializer.toStatic(doc, channelFields)
    }
    return snapshot
  }

  publish (channel, message) {
    const [channelName, channelId] = channel.split('.')
    this.db[channelName] = this.db[channelName] || []
    this.db[channelName].push({channelId, message})
  }
}
