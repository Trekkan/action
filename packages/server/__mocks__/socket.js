/* eslint-env jest */
const socket = {
  emit: jest.fn(),
  setAuthToken: jest.fn()
}

export default socket
