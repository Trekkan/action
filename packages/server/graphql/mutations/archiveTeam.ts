import {GraphQLID, GraphQLNonNull} from 'graphql'
import {SubscriptionChannel} from 'parabol-client/types/constEnums'
import getRethink from '../../database/rethinkDriver'
import NotificationTeamArchived from '../../database/types/NotificationTeamArchived'
import safeArchiveTeam from '../../safeMutations/safeArchiveTeam'
import {getUserId, isTeamLead} from '../../utils/authorization'
import publish from '../../utils/publish'
import segmentIo from '../../utils/segmentIo'
import standardError from '../../utils/standardError'
import ArchiveTeamPayload from '../types/ArchiveTeamPayload'

export default {
  type: GraphQLNonNull(ArchiveTeamPayload),
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The teamId to archive (or delete, if team is unused)'
    }
  },
  async resolve(_source, {teamId}, {authToken, dataLoader, socketId: mutatorId}) {
    const r = await getRethink()
    const operationId = dataLoader.share()
    const subOptions = {operationId, mutatorId}

    // AUTH
    const viewerId = getUserId(authToken)
    if (!(await isTeamLead(viewerId, teamId))) {
      return standardError(new Error('Not team lead'), {userId: viewerId})
    }

    // RESOLUTION
    segmentIo.track({
      userId: viewerId,
      event: 'Archive Team',
      properties: {
        teamId
      }
    })
    const {team, users, removedSuggestedActionIds} = await safeArchiveTeam(teamId)

    if (!team) {
      return standardError(new Error('Already archived team'), {userId: viewerId})
    }

    const notifications = users
      .map(({id}) => id)
      .filter((userId) => userId !== viewerId)
      .map(
        (notifiedUserId) =>
          new NotificationTeamArchived({userId: notifiedUserId, teamId, archivorUserId: viewerId})
      )

    if (notifications.length) {
      await r
        .table('Notification')
        .insert(notifications)
        .run()
    }

    const data = {
      teamId,
      notificationIds: notifications.map(({id}) => id),
      removedSuggestedActionIds
    }
    publish(SubscriptionChannel.TEAM, teamId, 'ArchiveTeamPayload', data, subOptions)

    users.forEach((user) => {
      const {id, tms} = user
      publish(SubscriptionChannel.NOTIFICATION, id, 'AuthTokenPayload', {tms})
    })

    return data
  }
}
