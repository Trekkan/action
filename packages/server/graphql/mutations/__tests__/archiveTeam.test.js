/* eslint-env jest */
import DynamicSerializer from 'dynamic-serializer'
import MockDate from 'mockdate'
import MockPubSub from '../../../__mocks__/MockPubSub'
import socket from '../../../__mocks__/socket'
import makeDataLoader from '../../../__tests__/setup/makeDataLoader'
import mockAuthToken from '../../../__tests__/setup/mockAuthToken'
import MockDB from '../../../__tests__/setup/MockDB'
import {__now} from '../../../__tests__/setup/mockTimes'
import fetchAndSerialize from '../../../__tests__/utils/fetchAndSerialize'
import getRethink from '../../../database/rethinkDriver'
import archiveTeam from '../archiveTeam'

MockDate.set(__now)
console.error = jest.fn()

describe('ArchiveTeam', () => {
  test('archives a team', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockPubSub = new MockPubSub()
    const mockDB = new MockDB()
    const {
      user,
      team: [updatedTeam],
      teamMember
    } = await mockDB.init()
    updatedTeam.isArchived = true
    const teamLeadId = teamMember.find((tm) => tm.teamId === updatedTeam.id && tm.isLead).userId
    const teamLead = user.find((usr) => usr.id === teamLeadId)
    const authToken = mockAuthToken(teamLead)
    const dataLoader = makeDataLoader(authToken)
    // TEST
    await archiveTeam.resolve(undefined, {teamId: updatedTeam.id}, {authToken, dataLoader, socket})

    // VERIFY
    const db = await fetchAndSerialize(
      {
        notification: r
          .table('Notification')
          .getAll(updatedTeam.orgId, {index: 'orgId'})
          .orderBy('startAt'),
        team: r.table('Team').get(updatedTeam.id)
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot()
  })
})
