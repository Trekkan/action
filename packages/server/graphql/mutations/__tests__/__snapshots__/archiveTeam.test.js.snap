// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchiveTeam archives a team 1`] = `
Object {
  "notification": Array [],
  "team": Object {
    "activeFacilitator": null,
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 0,
    "isArchived": true,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "name": "Team America",
    "orgId": 1,
    "tier": "personal",
  },
}
`;

exports[`ArchiveTeam archives a team 2`] = `
Object {
  "newAuthToken": Array [
    Object {
      "channelId": 2,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 3,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 4,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 5,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 6,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 7,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 8,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
    Object {
      "channelId": 9,
      "message": Object {
        "data": Object {
          "tms": Array [],
          "type": "updated",
        },
      },
    },
  ],
  "team": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "data": Object {
          "notificationIds": Array [
            10,
            11,
            12,
            13,
            14,
            15,
            16,
          ],
          "removedTeamNotifications": Array [],
          "teamId": 0,
          "type": "ArchiveTeamPayload",
        },
        "mutatorId": undefined,
        "operationId": 0,
      },
    },
  ],
}
`;
